### ISSUE TITLE
Champion "CallerArgumentExpression"

### ISSUE THREAD

**Issue by gafter** [dotnet/csharplang#287]
- [ ] Proposal added
- [ ] Discussed in LDM
- [ ] Decision in LDM
- [ ] Finalized (done, rejected, inactive)
- [ ] Spec'ed

See https://github.com/dotnet/csharplang/issues/205

**Comment by HaloFour** [comment]
If expansion of the supported caller info attributes is going to be discussed in the LDM is there a chance that some quantity of #87 be mentioned as well?

**Comment by CyrusNajmabadi** [comment]
@HaloFour I think that would be reasonable.  If we're going to do more in this space, we probably want to do it all at once.

**Comment by jamesqo** [comment]
Related corefx proposal that would take advantage of this feature: dotnet/corefx#17068

**Comment by jamesqo** [comment]
@CyrusNajmabadi 

> I think that would be reasonable. If we're going to do more in this space, we probably want to do it all at once.

Do I need to modify the proposal I'm working on to include his suggestions, or will they be reviewed just by looking at the issue?

**Comment by gafter** [comment]
My intent is to bring up this one request to the design group. It sounds like @CyrusNajmabadi will bring up the others in that discussion.

**Comment by alrz** [comment]
Related: https://github.com/dotnet/csharplang/issues/443 

```cs
void M(object param, [CallerArgumentExpression(nameof(param))] string arg = null) {}
```

**Comment by jamesqo** [comment]
@alrz I agree with the proposal, but it isn't strictly necessary for this to move forward. The end user will not have to touch this attribute. Only a handful of methods in the framework will use it, and for those it's fine to simply hard-code the string.

**Comment by jamesqo** [comment]
Great to see that this has been discussed in the LDM, I wish I'd commented just 2 days later so I would have seen that :smile: I see this is the text for the `CallerArgumentExpression` section:

> Push to 7.X and start process for getting the attribute in place.

Does that mean I should open a proposal in the corefx repo?

**Comment by jamesqo** [comment]
I opened a corefx proposal for this, see https://github.com/dotnet/corefx/issues/21809.

**Comment by paulomorgado** [comment]
Call me strange, but I still don't feel comfortable with optional arguments in public APIs.

To the proposal in question, the callee shouldn't be required to know anything about the caller. And the caller should know what it was doing just by the call stack of the exception. Otherwise, you have worst problems in your hands and this is not the solution.

**Comment by alrz** [comment]
Caller attributes are not a new concept so I think we should be past that argument by now.

**Comment by yaakov-h** [comment]
How would this work on extension methods?

I'm thinking specifically of something like [Shouldly](https://github.com/shouldly/shouldly) here.

**Comment by jamesqo** [comment]
@yaakov-h Oh wow, I hadn't even thought about this feature working in that way! I guess it'll do basically the same thing for extension methods-- capture everything before the dot that represents the expression is being invoked on. e.g. Off the top of my head:

```cs
public static class FooExtensions {
    public static void Bar(this Foo f, [CallerArgumentExpression("f")] string fExpression = null) {}
}

(this as Foo).Bar(); // fExpression = "(this as Foo)"
```

I have no idea how Shouldly works (I just read the README and it seems like dark magic), but it should be able to use CAE to get the string before the dot instead of whatever they use.

**Comment by yaakov-h** [comment]
@jamesqo it's definitely [dark magic](https://github.com/shouldly/shouldly/blob/master/src/Shouldly/Internals/SourceCodeTextGetter.cs#L38): it walks the stack, interprets the PDBs, then interprets the source code file. 😄 

Thanks for the spec update, looks reasonable to me. 👍

**Comment by alrz** [comment]
Debug.Assert can't take advantage of this feature, because there is already an overload that takes an string,
```cs
public static void Assert(bool condition);
public static void Assert(bool condition, string message);
```
If you make `message` an optional parameter, the compiler would still prefer the first overload.

I didn't come up with a solution to light up CallerArgumentExpression without losing source compatibility.

**Comment by jamesqo** [comment]
@alrz We can maintain source compatibility. However, we may have to break binary compatibility by removing the first overload.

**Comment by alrz** [comment]
Maybe you could do it in a library you use but that's simply not an option for Debug.Assert.

**Comment by paulomorgado** [comment]
There could be an helper type that takes all the values in the constructor as optional arguments:

```c#
public static void Assert(AssertContext context, bool condition);
public static void Assert(AssertContext context, bool condition, string message);
```

**Comment by alrz** [comment]
*Maybe* we could do this with a static extension method in "extention everything" feature,
```cs
namespace System.Diagnostics {
  public extension class DebugExtensions for Debug {
    public new static void Assert(bool condition,
      [CallerArgumetExpression("condition")] string message = default) { .. }
  }
}
```
If we were able to define `new` extension methods to hide the original, we're there.

**Comment by paulomorgado** [comment]
@alrz, are you expecting the extension method to win over the class method?

**Comment by alrz** [comment]
@paulomorgado If it's an `extension class`, yes, that comes naturally.

**Comment by alrz** [comment]
Note: I've defined it as `new`, this could be considered as part of "extension everything" feature.

**Comment by HaloFour** [comment]
@alrz

I don't see extension methods/classes supporting shadowing anymore than they would support virtual overrides.

**Comment by alrz** [comment]
@HaloFour 

virtual/override is out of the question. But when you are defining an extension class I think `new` could be useful actually, for this kind of use cases. 

Note: Currently `new` is only applicable to vertical extension i.e. inheritance. With `extension class` you are extending a type horizontally, with that in mind, `new` makes sense as much as virtual/override doesn't

**Comment by alrz** [comment]
I filed https://github.com/dotnet/roslyn/issues/20631 to show how an extention method can be useless if there already exists an instance method with the same name. While `new` doesn't make sense in that context, I think it actually does, when you are inside an `extension class`.

**Comment by jamesqo** [comment]
@alrz 

> Maybe you could do it in a library you use but that's simply not an option for Debug.Assert.

I think it can be an option. Let me explain my reasoning:

- `Debug.Assert` is stripped from Release builds, so the method won't be present in the IL for production builds. Only debug builds will be broken, which can be easily recompiled by the dev since presumably they're on the dev's machine.

- .NET Core seems to be a lot more agile than the .NET Framework. I realize the corefx team takes back compat very seriously, but this has such widespread and practical benefit that I believe this is a change worth making. Perhaps the API can be changed when the major version number is bumped (e.g. .NET Core 3.0 or 4.0) and at the same time other breaking changes are made, to soften the blow.

It will be quite a shame if `Debug` doesn't take advantage of this, as people will end up creating their own wrapper classes like this for every project:

```cs
public static class DebugEx
{
    public static void Assert(bool condition, [CAE("condition")] string message = null)
        => Debug.Assert(condition, message);
}

DebugEx.Assert(!_foo.IsEmpty);
```

/cc @karelz, do you want to chime in on this?

Just a reminder: whether or not this feature is eventually used by `Debug.Assert`, it still has much practical benefit in the way of better argument validation. (See the `Verify` class I wrote [in the proposal](https://github.com/dotnet/csharplang/blob/master/proposals/caller-argument-expression.md).)

**Comment by alrz** [comment]
>Debug.Assert is stripped from Release builds, so the method won't be present in the IL for production builds. Only debug builds will be broken, which can be easily recompiled by the dev since presumably they're on the dev's machine.

That's an interesting observation. Regardless, my claims about `new` still makes sense to me (admittedly unrelated here).

**Comment by jamesqo** [comment]
@alrz Interesting idea. However, your fix depends on a new language feature that's still in the planning phase. (Plus, I think it could be dangerous if third-party libraries were allowed to hide framework methods, but what I said earlier is my main issue.)

I think a simpler fix would be to introduce another `Debug` class in a new namespace, e.g. `System.Diagnostics.Extensions` as you said earlier. It would look like this:

```cs
namespace System.Diagnostics.Extensions
{
    public static class Debug
    {
        public static void Assert(bool condition, [CAE("condition")] string message = null)
            => System.Diagnostics.Debug.Assert(condition, messsage);
    }
}
```

Then people would simply insert this line into their usings:

```cs
using Debug = System.Diagnostics.Extensions.Debug;
```

And all usages of `S.D.Debug` would be replaced with `S.D.E.Debug`, which would take advantage of the new feature. [Here's](https://sharplab.io/#v2:D4AQDABCCMDcCwAocVoDoAiBLAhgcwDsB7AZwBcsBjEhZSDAUwCMBXPCAXlU10NIupoAogA8yDAiSxFJmZm1pICOALYMSABxyUG3bPmLkqJYWIlSZJCAG8kEe1ADMEcjgFQATBEat2txA6BTi5kblRQACwQAIIkJAwATmQAFExERAA2EJQyACYANKiQanH4ulwELBkZAJQ2EAC+dg5NiK1KqupaOhAAcjZIAJAgziBeAMIDAUH2I5EQALLJdf4zMz5saLHxSckAZjgZ8TW0a43N9q0NQA===) a demo on TryRoslyn to show that this indeed works.

**Comment by alrz** [comment]
@jamesqo

If and only if you've just used `Assert`. All other members will fail to resolve but I think it wouldn't be much of a problem.

**Comment by jamesqo** [comment]
@alrz Just add those methods too? I didn't say only `Assert`.

**Comment by alrz** [comment]
I don't like code duplication it introduce especially if you're proposing this for corefx, so I just striked it out from the get-go. But still could be a workaround, if you are willing to add the alias in every file in the project. Ideally this should work out-the-box otherwise there is a bunch of workarounds you could consider.

**Comment by jamesqo** [comment]
> Ideally this should work out-the-box otherwise there is a bunch of workarounds you could consider.

I like things working out-of-the-box too, which is why I said I'd prefer the binary-breaking change. Let's see what @karelz and @terrajobst think about this before we start discussing workarounds.

**Comment by jcouv** [comment]
The email discussion with LDM is still going on, but a new issue was raised: how do we expect existing APIs (such as `Debug.Assert` or XUnit's `Assert.True`) to adopt this attribute?

**Comment by jamesqo** [comment]
@jcouv I don't think that there's a way to get around this without breaking binary compatibility. However, the attribute is mainly for diagnostic purposes (debug assertions, unit tests). 1) I don't think breaking unit tests is as important as breaking non-test code (@bradwilson, please chime in here or at xunit/xunit#1596). 2) As for `Debug.Assert`, changing the signature will only break code compiled in Debug mode, which is typically on the developer's machine, so (s)he can just recompile the code to fix the binary compat issue. So IMO, it is worth breaking debug code for this. The framework team should voice their thoughts though.

There's another point I want to raise. As I mention in the xunit issue above, one `string` parameter would need to be added for each existing parameter if we want to report as much info as possible, so method signatures could get pretty hairy. (e.g. `Equal(T, T)` would become `Equal(T, T, string, string)`)

In spite of being the one who made this proposal in the first place, I think this seems a little fragile. What if later on we add another attribute that exposes caller information, and we decide we want to display that info as well on test failures? Then more optional parameters would have to be added, and binary compat would need to be broken again.

There's a solution to this problem: we could enscapulate all the caller info in a struct, as @HaloFour suggested previously. Then we would need to break binary compat only once. [I outlined this solution at the bottom of my proposal](https://github.com/dotnet/csharplang/blob/master/proposals/caller-argument-expression.md#alternatives) (second bullet point). What are people's thoughts on that?

/cc @bradwilson @JosephWoodward

**Comment by bradwilson** [comment]
> I don't think breaking unit tests is as important as breaking non-test code

In the case of `xunit.assert.dll`, I am fine with binary breaking changes that aren't compile-time breaking changes. The class is `static`, the methods are `static`, thus they cannot be overridden. Nobody should be shipping a new assertion library that's written on top of it, either; they should be using it as a Git submodule (and ideally hiding the old API, so there aren't two of them).

> As I mention in the xunit issue above, one string parameter would need to be added for each existing parameter if we want to report as much info as possible, so method signatures could get pretty hairy.

I'm not 100% sure that I would want that, for the same reasons that we don't let people pass user messages today (except on True and False, as the escape hatches). We have philosophical objection to writing "user messages" instead of better asserts.

> In spite of being the one who made this proposal in the first place, I think this seems a little fragile.

Put me down as "I'm not a fan of the current design".

> There's a solution to this problem: we could enscapulate all the caller info in a struct

I'm not a big fan of that design, either, but it's probably better. Having intrinsic values (strings, ints, what-have-you) which are still shown as user-callable in the signature is sort of the worst scenario, IMO. Users should NEVER be able to provide values for these things.

I don't have any alternative to offer.

(xUnit.net has an additional problem which is: we would not be willing to take new platform dependencies to get this feature. Unless it's implemented purely contractually via the compiler, we will likely never support it until such time as we were willing to throw all the old platforms away. We're not there yet. 😉)

**Comment by gafter** [comment]
Would this hypothetical struct contain the string representation of all arguments in a map indexed by parameter name? If not, how would it address this use case?

-Neal

On Dec 20, 2017, at 7:12 PM, James Ko <notifications@github.com<mailto:notifications@github.com>> wrote:


@jcouv<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fjcouv&data=02%7C01%7Cneal.gafter%40microsoft.com%7C00e3bd9be8fa48d33aca08d54820aac9%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636494227566135878&sdata=p9XYzSgOdKmZFzv%2Bn4bq2hKLeZQupg851GC4uuejAO8%3D&reserved=0> I don't think that there's a way to get around this without breaking binary compatibility. However, the attribute is mainly for diagnostic purposes (debug assertions, unit tests). 1) I don't think breaking unit tests is as important as breaking non-test code (@bradwilson<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fbradwilson&data=02%7C01%7Cneal.gafter%40microsoft.com%7C00e3bd9be8fa48d33aca08d54820aac9%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636494227566135878&sdata=qSLI9Q4u%2FU%2F2d6md2WnRFmHXy9gZdbxxigu5ENzoZkM%3D&reserved=0>, please chime in here or at xunit/xunit#1596<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fxunit%2Fxunit%2Fissues%2F1596&data=02%7C01%7Cneal.gafter%40microsoft.com%7C00e3bd9be8fa48d33aca08d54820aac9%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636494227566135878&sdata=1ic43kDi3zz%2B%2F%2FwX05M21RtLq5iu9CVaHnuXgObDOoE%3D&reserved=0>). 2) As for Debug.Assert, changing the signature will only break code compiled in Debug mode, which is typically on the developer's machine, so (s)he can just recompile the code to fix the binary compat issue. So IMO, it is worth breaking debug code for this. The framework team should voice their thoughts though.

There's another point I want to raise. As I mention in the xunit issue above, one string parameter would need to be added for each existing parameter if we want to report as much info as possible, so method signatures could get pretty hairy. (e.g. Equal(T, T) would become Equal(T, T, string, string))

In spite of being the one who made this proposal in the first place, I think this seems a little fragile. What if later on we add another attribute that exposes caller information, and we decide we want to display that info as well on test failures? Then more optional parameters would have to be added, and binary compat would need to be broken again.

There's a solution to this problem: we could enscapulate all the caller info in a struct, as @HaloFour<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fhalofour&data=02%7C01%7Cneal.gafter%40microsoft.com%7C00e3bd9be8fa48d33aca08d54820aac9%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636494227566135878&sdata=QV0nVP%2FoPMi8gQO1wYGQ7eR0Grs%2B8m7U0EWbYZTGUmQ%3D&reserved=0> suggested previously. Then we would need to break binary compat only once. I outlined this solution at the bottom of my proposal<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fdotnet%2Fcsharplang%2Fblob%2Fmaster%2Fproposals%2Fcaller-argument-expression.md%23alternatives&data=02%7C01%7Cneal.gafter%40microsoft.com%7C00e3bd9be8fa48d33aca08d54820aac9%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636494227566135878&sdata=p991nnm%2BsXyZEQTTXN1M9pSi0T2oq2YkDQwbZUfrMIc%3D&reserved=0> (second bullet point). What are people's thoughts on that?

/cc @bradwilson<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fbradwilson&data=02%7C01%7Cneal.gafter%40microsoft.com%7C00e3bd9be8fa48d33aca08d54820aac9%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636494227566135878&sdata=qSLI9Q4u%2FU%2F2d6md2WnRFmHXy9gZdbxxigu5ENzoZkM%3D&reserved=0> @JosephWoodward<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fjosephwoodward&data=02%7C01%7Cneal.gafter%40microsoft.com%7C00e3bd9be8fa48d33aca08d54820aac9%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636494227566135878&sdata=Pkj8RcsGqMofd2tQtQmZBHfAXCgkUPAsKxSX27P5pzI%3D&reserved=0>

**Comment by bradwilson** [comment]
It would have to contain a dictionary of argument names to expressions, in addition to whatever other caller attributes they choose to port/add (I think today they support method name, filename, and line number, but there has been talk of adding other things as well).

**Comment by jamesqo** [comment]
@gafter @bradwilson No it would not. It would simply contain a string array. The 0th index would contain the expression for the 1st parameter name, 1st index would correspond to the 2nd parameter name, etc. This approach would allow us to support `params` arguments.

**Comment by alrz** [comment]
> Users should NEVER be able to provide values for these things.

I can imagine an analyzer could flag those cases. But in case of CallerArgumnetExpression it's actually useful?


>Would this hypothetical struct contain the string representation of all arguments

I think it would be useful to define a contract for user to define their own CallerInfo to avoid passing unnecessary data. We may have a reference impl that contains all caller infos.

**Comment by alrz** [comment]
As for overload resolution, we can (ab)use the base class name lookup,

First we move the existing Debug members to base and introduce new members in a derived class.
```cs
namespace System.Diagnostics {
  public class DebugBase
  {
    // existing members
    public static void Assert(bool c, string m) {}
    public static void Assert(bool c) {}    
  }
    
  public class Debug : DebugBase {
    public static new void Assert(bool c, [CallerArgumnetExpression("c")] string m = null) {}	
  }  
}
```
Now `Debug.Assert(false)` calls the new member: `Debug.Assert(false, "false");`

I suspect `call Debug.Assert(bool)` can be resolved to `call DebugBase.Assert(bool)` at runtime.

**Comment by paulomorgado** [comment]
@jamesqo, and, how about [Trace.Assert](https://msdn.microsoft.com/library/system.diagnostics.trace.assert "Trace.Assert Method")?

**Comment by AntonPlotnikov** [comment]
Is there any progress in this proposal since 2017?

**Comment by jamesqo** [comment]
@AntonPlotnikov Its been approved, up-for-grabs here: https://github.com/dotnet/corefx/issues/21809

**Comment by jcdickinson** [comment]
> @jcouv how do we expect existing APIs (such as Debug.Assert or XUnit's Assert.True) to adopt this attribute?

Allow it on non-optional parameters. The Debug.Assert case would become:

```c#
namespace System.Diagnostics
{
    public static class Debug
    {
        public static void Assert(bool condition)
            => Assert(condition, null);

        public static void Assert(bool condition, [CAE(nameof(condition))] string message)
            => // ...
    }
}
```

The compiler should prefer CAE. If the developer doesn't want to use CAE, they can pass `null` or whatever the original default value would have been.

Alternatively, don't use `string`. The attribute says "Expression," so use `Expression`:

```c#
namespace System.Diagnostics
{
    public static class Debug
    {
        public static void Assert(bool condition) => // ...

        public static void Assert(bool condition, string message) => // ...

        public static void Assert(bool condition, [CAE(nameof(condition))] Expression accessor = null)
            => // ...
    }
}
```

Stringifying these might be expensive, but remember that these will typically be used in debug (unit tests) or exceptional (contracts) cases.

**Comment by jcouv** [comment]
@jcdickinson If you're proposing to adjust overload resolution rules to make this work more smoothly, I think it's very unlikely to happen. The cost/benefit is not worth it and it would be very hard to rationalize in the language design (consider more complex scenarios where multiple candidate overload have CAE, etc).

By current overload resolution rules, if you call `Assert(true)` with the overloads you described, the one with a single `bool` parameter [wins](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxAgrgOyQExAagB8ABABgAJiBGAbgFgAoYgZkoCZyBhcgb0fIGVW1AGyUALJSoAOABQBKcv0F8Gg9eQAicYOgDmAOgCCUKIhiyYGOPPprBAX0ZO1LKWOIdtuvcoGqNQTdRCXITMwQLYAB7aIAbcgBjaMwcAEsYNJTFHikDThSoeLgDAHUEDLhZACIYAAs0qHJogDdEOOiIHGrbchc/DWCqD0lw81kY+KSU9MyUpCkKMDhTCD0bcgBeAD5yeoRogHdyTHQ4uLsB9SGRsNNxyYTk1Iys7HIAUQAPAAcEFagb3IEESiQB0QQWxOZziih2ezqB2Op3OdnU/SYwg43z+AKBuQcQA).

**Comment by jswolf19** [comment]
@jamesqo I'm working on adding this attribute to corefx. I want to eventually add some tests for after the feature is added to let people know what to expect, and I have a couple of questions about the expected behavior.

1. For extension methods for use with the this parameter, you pass "this" to the attribute. Is there a reason to prefer that over the parameter name? 
`public static void ShouldBe<T>(this T actual, T expected, [CallerArgumentExpression("this")] string thisExpression = null)`
vs
`public static void ShouldBe<T>(this T actual, T expected, [CallerArgumentExpression("actual")] string thisExpression = null) {}`

2. (Possibly related to the above) do you feel that "this" should be supported for instance methods to get the this expression?

3. How should tokens that the compiler ignores be handled? (whitespace, inline comments, etc?)

I have some examples for these and other cases, such as parameters with default values over at [corefx issue 21809](https://github.com/dotnet/corefx/issues/21809#issuecomment-394165836), if you or others could provide some feedback, I'd appreciate it.

Thanks ^_^

**Comment by jamesqo** [comment]
@jswolf19 Sorry for the delayed reply!

1. The second snippet is correct, it should be `[CAE("actual")]` to get the `this` expression.

2. Not at all I guess? People can just make an extension method if they need that functionality. @jcouv comment if you agree/disagree.

3. Whitespace, inline comments should not be included in the expression. We don't want people having to call `.Trim()` and/or some hacky stuff to detect and remove comments before displaying the expression to the user.

**Comment by jswolf19** [comment]
@jamesqo Thanks for your reply. Since you replied at [corefx21809](https://github.com/dotnet/corefx/issues/21809#issuecomment-399626324), I'll continue the conversation there.

**Comment by CyrusNajmabadi** [comment]
> Leading/trailing whitespace, inline comments should not be included in the expression. We don't want people having to call .Trim() and/or some hacky stuff to detect and remove comments before displaying the expression to the user.

Why would someone need to "detect and remove comments before displaying to the expression to the user".  The expression is the expression.  Why not show the user exactly what was typed?

--

If we are going to start futzing with the expression are we also going to normalize whitespace in it?  Remove newlines?  Deal with preprocessor trivia? etc. etc.?  It seems pretty wonky, and means a lot of code to main and enforce (especially as the language changes).  

I would be much happier to just say: get the span of the argument from the beginning of the first token to the end of the last token.  Grab the text for that span.  Pass it along.

Simple, clean, and no need to futz around with this.  This also will mean that people can use this feature *and* use comments to pass along info they may want to pass along into deeper parts of the system.  

--

Here's an example of how this coudl bite us badly.  There are ideas around some sort 'attribute anywhere' in C#.  This would likely be represented in our system as some form of structured-trivia (like we represent preprocessor directives).  We would not have to decide if we wanted to keep this sort of info.  Even though it would be 'compiler ignored' (since it would only be for analyzers and other systems to use), it could definitely be something someone would want to keep along for use in a CAE and not lose.  

--

IMO, we should just keep it all.  It's simple, easy, and doesn't block any future scenarios.

**Comment by michal-ciechan** [comment]
> Also, don't use `string`. The attribute says "Expression," so use `Expression`:

Personally I prefer a `string`.

But it could always work with both? E.g. 

```csharp

void Assert(bool condition, [CAE("condition")] Expression argument = null)

// Provide string expression as it appears in source
void Assert(bool condition, [CAE("condition")] string argument = null)
```

I can definitely see value in both, but I imagine with the `Expression`, you get a whole load of issues of potentially either needing to capture variables, or provide palceholders to reason about.

If performance isn't an issue, you can just pass in the "Expression", compile and run it to get the condition, as a number of libraries already do (`Moq` and `FluentAssertions` come to mind).

> IMO, we should just keep it all. It's simple, easy, and doesn't block any future scenarios.

I definitely agree with this, I would 100% prefer to keep it all! I much prefer to throw the string at a expression parser after if I really need it pretty printed (or specially formatted), do so myself, as format and what to keep/discard is very opinionated

**Comment by jswolf19** [comment]
@michal-ciechan I went with string for the tests, because the [proposal](/dotnet/csharplang/blob/9d5989ce2dbf1bf462543151e4240fb349e72441/proposals/caller-argument-expression.md) defines it as "an attribute the framework can use to obtain the string associated with a particular method argument."

As for the tests and the current expected results, I added tests more for eventual documentation on how the feature works for different cases. While I tried to provide expecteds that what the proposers were imagining, I also thought that the expected result could/should be fixed to match the implementation rather than necessarily making the implementation match the tests in their current form (thus my comments on the currently removed `Skip` attributes that "Final behavior may differ.")

As for where it goes from here, I'm still very interested in seeing it in the wild ^^

**Comment by SamPruden** [comment]
> IMO, we should just keep it all. It's simple, easy, and doesn't block any future scenarios.

Is there any reason not to provide some common transformations as opt-ins toggled with additional attribute arguments? If any of them are a bit expensive or fiddly (like removing comments) it might be better to have them done once at compile time rather than every time at runtime. Keeping it all would be the default.

I think `.Trim()` is a rather obvious transformation that would very commonly be wanted. Some people write `Debug.Assert(a > b);` and some write `Debug.Assert( a > b );` but I expect that a typical assertion framework would want to normalize these in the output.

**Comment by abelbraaksma** [comment]
I'm a little confused as to the status of this. In the original post above it says that there's no decision yet and it isn't finalized yet, let alone an implementation. However, [the documentation](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.callerargumentexpressionattribute?view=netcore-3.1) suggests that this has been available since .NET Core 3.0.

A [nice blog post on CallerArgumentExpressionAttribute](https://blog.mcilreavy.com/articles/2018-08/caller-argument-expression-attribute) from 2018 suggests it's up and coming.

The attribute is available in the .NET Core libraries, but given this issue is still open I wonder if, and if so, by how much, it's actually functional right now.

**Comment by 333fred** [comment]
The attribute exists, but compiler support is not yet implemented, though hopefully it will be soon.

**Comment by sschultze-tb** [comment]
> The attribute exists, but compiler support is not yet implemented, though hopefully it will be soon.

I also spent quite some time trying to get this to work before I read your post. 😫 The CallerArgumentExpressionAttribute documentation should really be updated with a note that this is for future use only and not functional right now. 😄

**Comment by dominikjeske** [comment]
I’m also confused – this attribute CallerArgumentExpressionAttribute Class (System.Runtime.CompilerServices) | Microsoft Docs should not be in docs if it is not working.

**Comment by AlexFree** [comment]
Can it be supported also in .net standard 2.1 ?

**Comment by chucker** [comment]
> Can it be supported also in .net standard 2.1 ?

Someone can make a source-only NuGet to retrofit the argument.

**Comment by abelbraaksma** [comment]
> Can it be supported also in .net standard 2.1 ?

It isn't even supported in 3 or 5. The docs are wrong.

**Comment by NetMage** [comment]
@ghost I submitted a suggest documentation change.

**Comment by MichaelPeter** [comment]
Is there an update for it? Would be REALLY usefull for my Guard classes, like also described here by Paul Mcilreavy. That was 2018!

https://blog.mcilreavy.com/articles/2018-08/caller-argument-expression-attribute

**Comment by 333fred** [comment]
> Saw there have been pull requests in March...

That's the update: implementation work has started, and is being driven by a community member.

**Comment by benfoster** [comment]
Are there plans to provide a shim package for .NET Standard?

**Comment by 333fred** [comment]
> Are there plans to provide a shim package for .NET Standard?

Not AFAIK.

**Comment by jnm2** [comment]
I prefer internal attribute class declarations over shim packages, myself.

**Comment by vbcodec** [comment]
simpler version, without unnecessary clutter
```
    void fun (int arg)
    {
        string expr = GetCallingExpression(arg);
    }
```
where `GetCallingExpression` is build-in function that return stringified expression and can be called only with function's parameters.

**Comment by jnm2** [comment]
@vbcodec That signature would be misleading, since the actually emitted method must have an additional string parameter. There's no other general mechanism for obtaining the string value than having the string value passed via parameter.

**Comment by vbcodec** [comment]
> 
> 
> There's no other general mechanism for obtaining the string value than having the string value passed via parameter.

It is matter of compiler. Compiler can generate such additional parameters, or use thread-based ambient context (must be added to NET 6 BCL), to transfer these strings

**Comment by CyrusNajmabadi** [comment]
> It is matter of compiler. 

It can't.  Source code doesn't exist at runtime.  

>  or use thread-based ambient context (must be added to NET 6 BCL), to transfer these strings

There's no thread based ambient context that contains the original source text to transfer.

**Comment by IanKemp** [comment]
For future visitors: https://docs.microsoft.com/dotnet/api/system.runtime.compilerservices.callerargumentexpressionattribute

**Comment by jnm2** [comment]
> Since .NET Standard is dead, sadly this attribute will never appear there.

@IanKemp This actually doesn't matter. .NET Standard projects can do this:

```cs
#if NETSTANDARD
namespace System.Runtime.CompilerServices;

[AttributeUsage(AttributeTargets.Parameter, AllowMultiple = false, Inherited = false)]
internal sealed class CallerArgumentExpressionAttribute : Attribute
{
    public CallerArgumentExpressionAttribute(string parameterName)
    {
        ParameterName = parameterName;
    }

    public string ParameterName { get; }
}
#endif
```

This general approach works for all compiler-recognized attributes. I first used it with CallerMemberNameAttribute, and SerializableAttribute was another example of this before it was added to .NET Standard.

It's very important to declare these attributes as internal so that consuming projects don't end up seeing conflicting definitions of System.Runtime.CompilerServices.CallerArgumentExpressionAttribute. Declaring the attribute as internal does not restrict where it can be used within the project that declared it.

**Comment by NetMage** [comment]
That is no help in gaining the actual functionality of the attribute, which means it has no purpose if you are targeting .Net Standard.

**Comment by jnm2** [comment]
@NetMage That's incorrect. The actual functionality of the attribute is in full effect so long as you are using the .NET 6 SDK. The same thing has been true of all compiler-recognized attributes; the target framework doesn't affect compiler behavior so long as somebody declares the correct types.

**Comment by jnm2** [comment]
You can demonstrate for yourself that this works when targeting .NET Standard. Here's one way:

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFrameworks>netstandard2.0</TargetFrameworks>
    <LangVersion>10</LangVersion>
  </PropertyGroup>

</Project>
```

##### Program.cs

```cs
using System;
using System.Runtime.CompilerServices;

WriteExpression(1 + 2);

static void WriteExpression(object p, [CallerArgumentExpression("p")] string expression = null)
{
    Console.WriteLine(expression);
}
```

##### System.Runtime.CompilerServices.CallerArgumentExpressionAttribute.cs

```cs
#if !NETCOREAPP3_0_OR_GREATER // This line is only needed if you are multitargeting with netcoreapp3.0+
namespace System.Runtime.CompilerServices;

[AttributeUsage(AttributeTargets.Parameter, Inherited = false)]
internal sealed class CallerArgumentExpressionAttribute : Attribute
{
    public CallerArgumentExpressionAttribute(string parameterName)
    {
        ParameterName = parameterName;
    }

    public string ParameterName { get; }
}
#endif
```

Result:

![image](https://user-images.githubusercontent.com/8040367/141661544-157f90f6-bc31-4304-bc96-a559dfcc4e14.png)

**Comment by bradwilson** [comment]
I can verify this works with .NET 4.7.2, .NET Core 3.1, and .NET Standard 2.0 given the example here. I've already updated my `Guard` class in xUnit.net: https://github.com/xunit/xunit/commit/d53d0761f997ae132b9147c2347659029c9d0e23

Thanks! 🎉

**Comment by dominikjeske** [comment]
Why is this issue still open? Something missing?

**Comment by bernd5** [comment]
ECMA Spec is missing

**Comment by 333fred** [comment]
We keep issues open until they are documented in the ECMA specification over at dotnet/csharpstandard.

**Comment by abelbraaksma** [comment]
That's now 3 years later, is the ECMA spec that far behind, or was this simply forgotten? Where should this go? I don't mind giving it a go, as I'm looking at implementing this for F# (see https://github.com/fsharp/fslang-suggestions/issues/966).

**Comment by colejohnson66** [comment]
The 2023 ECMA is for C# 7. Yes, it’s the far behind. Progress is being made at @dotnet/csharplang. They’re working on C# 8 ATM.
